src/QuickCheckVEngine/RVFI_DII/DII.hs

```diff
@@ -115,3 +117,15 @@ diiEnd :: DII_Packet
 diiEnd = DII_Packet { dii_cmd  = dii_cmd_end
                     , dii_time = 1
                     , dii_insn = 0 }
+
+-- | Construct a version negotiation 'DII_Packet'
+diiVersNegotiate :: DII_Packet
+diiVersNegotiate = DII_Packet { dii_cmd  = dii_cmd_end
+                    , dii_time = 1
+                    , dii_insn = 0x56455253 } -- send 'V' 'E' 'R' 'S'
+
+-- | Construct a version request 'DII_Packet'
+diiRequestVers :: Word32 -> DII_Packet
+diiRequestVers vers = DII_Packet { dii_cmd  = 0x76 -- 'v'
+                                  , dii_time = 1
+                                  , dii_insn = vers }
```

Comment:
Isn't it nicer to just fully create 2 new `DII_Cmd` on top of `dii_cmd_instruction` and dii_cmd_end`?
(say `dii_cmd_vers_negotiate` and `dii_cmd_request_vers`

Reply:
I'm afraid I don't understand what you want me do here.

Reply:
Sorry that was very poorly phrased. What I mean is so far had 2 sorts of packets: instruction packets and end packets. Really, the `dii_cmd` field identified which sort it was. If I understand correctly, there is now notionally two new sorts of packets: version negotiation and version request packets.

I am suggesting they get their own `dii_cmd` value, and for those to be defined up at the top of the file next to `dii_cmd_instruction` and `dii_cmd_end`. Does it make sense?
(the version negitiation packet seam to reuse the `dii_cmd_end` together with a magic value in `dii_insn` which I presune could work instead just as well with a dedicated `dii_cmd` value and be more readable? And the version request defines an ad-hoc value for the `dii_cmd` inline as a literal straight in the packet, so this as well would benefit from being listed next to the other `dii_cmd` I think...)

Reply:
The whole point of reusing `dii_cmd_end` with a special `dii_insn` is so that existing v1 RVFI-DII implementations continue to work as they'll just ignore the `dii_insn` field but v2 implementations know they can send a version information packet. Although it seems like we have one more version negotiation packet than needed? Currently there's the version negotiation packet, then the versioned halt reply, then the version request packet? It _should_ only need one round trip unless I'm missing something.

Also Alex it would probably be wise to have a way to express the _minimum_ supported version too so we can eventually transition away from v1.

Reply:
Right I forgot about the "we don't touch existing implementations" bit, sure. So at least the version request one should have an explicit `dii_cmd` up there (that is if still deemed useful after @jrtc27 's comment).

Reply:
We have 87 effectively unused bytes in the initial halt packet so we could report a lot more data.

It would be possible to drop the second packet, however I was thinking of the following sequences:
```
vengine:: reset with 'VERS'
impl: I support v2
vengine: that's great but I don't (not sure if we need to support old vengine) -> continue with v1
```

```
vengine:: reset with 'VERS'
impl: I support v2
vengine: give me version 2 traces please
impl: okay sending v2 from now on
```

```
vengine:: reset with 'VERS'
impl: I support v2
vengine: I don't care, give me version 1 traces please
Either impl: okay sending v1 from now on
Or impl: no I can only send v2
vegine: got a version=2 response -> vengine can accept unexpected version or fatal error
```

I agree that a min_version/selected_version in the initial packet would simplify that and if min_version matches what the vengine wants we can skip the set-version command.
However, @gameboo was uncomfortable with reusing fields that have a defined meaning in the RVFI spec. This implementation just modifies the padding bits of rvfi_halt to report "I can do more than just v1 traces", but I guess we could the padding in the other one-bit fields such as rvfi_trap, etc. to report selected_ver/min_ver.